src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,596 @@
+
+        for endpoint_id, endpoint in self.endpoints.items():
+            for cluster_type, cluster in endpoint.items():
+                if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                    continue
+
+                all_clusters = set(list(ClusterObjects.ALL_CLUSTERS.keys()))
+                dut_clusters = set(list(x.id for x in endpoint.keys()))
+
+                unsupported = [id for id in list(all_clusters - dut_clusters) if global_attribute_ids.attribute_id_type(id)
+                               == global_attribute_ids.AttributeIdType.kStandardNonGlobal]
+
+                if unsupported:
+                    unsupported_attribute = (ClusterObjects.ALL_ATTRIBUTES[unsupported[0]])[0]
+
+                    result = await self.read_single_attribute_expect_error(endpoint=endpoint_id, cluster=ClusterObjects.ALL_CLUSTERS[unsupported[0]], attribute=unsupported_attribute, error=Status.UnsupportedCluster)
+                    asserts.assert_true(isinstance(result.Reason, InteractionModelError),
+                                        msg="Unexpected success reading invalid cluster")
+
+        # Step 21
+        # TH sends the Read Request Message to the DUT to read an unsupported attribute
+        # DUT responds with the report data action.
+        # Verify on the TH that the DUT sends the status code UNSUPPORTED_ATTRIBUTE
+        self.print_step(21, "Send the Read Request Message to the DUT to read any attribute to an unsupported attribute")
+        found_unsupported = False
+        for endpoint_id, endpoint in self.endpoints.items():
+
+            if found_unsupported:
+                break
+            for cluster_type, cluster in endpoint.items():
+                if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                    continue
+
+                all_attrs = set(list(ClusterObjects.ALL_ATTRIBUTES[cluster_type.id].keys()))
+                dut_attrs = set(cluster[cluster_type.Attributes.AttributeList])
+
+                unsupported = [id for id in list(all_attrs - dut_attrs) if global_attribute_ids.attribute_id_type(id)
+                               == global_attribute_ids.AttributeIdType.kStandardNonGlobal]
+
+                if unsupported:
+                    result = await self.read_single_attribute_expect_error(
+                        endpoint=endpoint_id,
+                        cluster=cluster_type,
+                        attribute=ClusterObjects.ALL_ATTRIBUTES[cluster_type.id][unsupported[0]], error=Status.UnsupportedAttribute)
+                    asserts.assert_true(isinstance(result.Reason, InteractionModelError),
+                                        msg="Unexpected success reading invalid attribute")
+                    found_unsupported = True
+                    break
```

Comment:
This will only break you out of this cluster loop - you'll still run this once per endpoint. suggest just popping this into a function and returning.

Reply:
Moved to a new function